### PR TITLE
Dotnet.exe push now has --skip-duplicate

### LINKED ISSUE #8100: `dotnet.exe nuget push` command now supports -skipDuplicate like nuget.exe did.
## Details about Problem
Related Issue: https://github.com/NuGet/Home/issues/1630

In the related issue, NuGet.exe push was given a new option, -SkipDuplicate. This option needs to be made available in DotNet.exe nuget push as --skip-duplicate.

NuGet product used: dotnet.exe

### PR DESCRIPTION
**Pull request by donnie-msft**
## Bug

Fixes: https://github.com/NuGet/Home/issues/8100

## Fix

Details: 

- Adds the `--skip-duplicate` option to dotnet.exe.

- Advertising the option when a 409-Conflict HTTP response has been enabled for dotnet.exe (already in nuget.exe).

- Advertising message was modified to show the name of the option:
nuget.exe: `To skip already published packages, use the option -SkipDuplicate`
dotnet.exe: `To skip already published packages, use the option --skip-duplicate`

- Added some XPlat FuncTests for testing server's with real response codes, but these are disabled for now (see below).

- Minor: refactored NuGet.exe FuncTests to use the "skip duplicate" terminology.

## Testing/Validation

Tests Added: Yes (but disabled) 
Reason for not adding tests:  
The core logic is shared with nuget.exe and is already being tested by the mock server testing in the client [PushCommandTest](https://github.com/NuGet/NuGet.Client/blob/3e3a03c9a7f65a51ca30a3a46e7571f5ef919279/test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs#L233).
We do not currently have any test servers setup to do automated testing for HTTP responses. These tests have been skipped for a long while. I improved the Skip messages for the PackageSources that were there, and also improved the test in general so it's in a better place when we do reintroduce this test. Issue to correct this has been created: https://github.com/NuGet/Home/issues/8695

We will also add this to our Manual Testing process.

Validation:  
**dotnet nuget push** with and without the --skip-duplicate option. Verified push continued with the option specified. Verified advertisement message showed and with correct option for dotnet. Verified Help showed correct message.

**nuget.exe push** made comparable verifications (for -skipduplicate) to make sure there were no regressions.

### CODE REVIEW

File: src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs [hunk 1]
```diff
@@ -452,21 +450,24 @@ private static bool DetectAndLogSkippedErrorOccurrence(HttpStatusCode? skippedEr
...
         {
-            
             string advertiseDescription = null;
 
             switch (errorCodeThatOccurred)
             {
                 case HttpStatusCode.Conflict:
-                    advertiseDescription = Strings.PushCommandSkipDuplicateAdvertise;
+
+                #if IS_DESKTOP
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Resources/PackageUpdateResource.cs]: style: Don't indent preprocessor definitions like #if.

File: src/NuGet.Core/NuGet.Protocol/Strings.resx [hunk 2]
```diff
@@ -519,8 +519,8 @@ The "ms" should be localized to the abbreviation for milliseconds.</comment>
     <value>Problem starting the plugin '{0}'. {1}</value>
     <comment>0 - plugin path, 1 - error message </comment>
   </data>
-  <data name="PushCommandSkipDuplicateAdvertise" xml:space="preserve">
-    <value>See help for push option to automatically skip duplicates.</value>
+  <data name="PushCommandSkipDuplicateAdvertiseNuGetExe" xml:space="preserve">
+    <value>To skip already published packages, use &lt;command&gt; -skipduplicate</value>
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Strings.resx]: Isn't the command -SkipDuplicate? 
Use the same casing.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs [hunk 3]
```diff
@@ -18,7 +18,7 @@ public class PushCommandTest
         private const string MESSAGE_PACKAGE_PUSHED = "Your package was pushed.";
         private const string TEST_PACKAGE_SHOULD_NOT_PUSH = "The package should not have been pushed";
         private const string TEST_PACKAGE_SHOULD_PUSH = "The package should have been pushed";
-        private const string ADVERTISE_CONTINUE_OPTION = "See help for push option to automatically skip duplicates.";
+        private const string ADVERTISE_SKIPDUPLICATE_OPTION = "To skip already published packages, use <command> -skipduplicate"; //PushCommandSkipDuplicateAdvertiseNuGetExe
```
Comment (nkolev92) [inline test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: There's an internals visible to from nuget.commandline to this assembly. 
You can just use the resource directly.

File: test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatPushTests.cs [hunk 4]
```diff
@@ -85,26 +87,35 @@ public async Task PushToServerSkipDuplicateSucceeds(PackageSource packageSource)
...
-
-                //info: PUT http://localhost:5000/api/v2/package/
-                //info: Conflict http://localhost:5000/api/v2/package/ 127ms
-                 Assert.Contains("already exists at feed", outputMessages);
-                Assert.Contains("Your package was pushed.", outputMessages);
+                var exitCodeFirstPush = NuGet.CommandLine.XPlat.Program.MainInternal(pushArgs.ToArray(), logFirstPush);
+                var exitCodeSecondPush = NuGet.CommandLine.XPlat.Program.MainInternal(pushArgs.ToArray(), logSecondPush);
 
+                // Assert First Push - it should happen without error.
+                Console.WriteLine("Waiting for debugger to attach.");
```
Comment (nkolev92) [inline test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatPushTests.cs]: Remove the next 5 lines.

File: test/TestUtilities/Test.Utility/PlatformXunitAttributes/PackageSourceTheoryAttribute.cs [hunk 5]
```diff
@@ -28,7 +28,7 @@ public sealed class PackageSourceTheoryAttribute : TheoryAttribute
         public override string Skip
         {
             get
-            {
+            {
```
Comment (nkolev92) [inline test/TestUtilities/Test.Utility/PlatformXunitAttributes/PackageSourceTheoryAttribute.cs]: Unnecessary change.

**Review by nkolev92 — COMMENTED** [review]
Mostly cosmetic suggestions.

File: src/NuGet.Core/NuGet.Protocol/Strings.resx (hunk 2 above)
Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Protocol/Strings.resx]: It's actually case-insensitive

Comment (zivkan) [reply src/NuGet.Core/NuGet.Protocol/Strings.resx]: I think the point is that our messages should be consistent. Use the same casing in docs, warning messages, help messages, etc. Using Pascal Casing makes it easier to read than all lowercase.

File: test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs (hunk 3 above)
Comment (donnie-msft) [reply test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/PushCommandTest.cs]: This is the Protocol assembly

**Review by nkolev92 — APPROVED** [review]

File: src/NuGet.Core/NuGet.Protocol/Strings.resx [hunk 6]
```diff
@@ -519,8 +519,8 @@ The "ms" should be localized to the abbreviation for milliseconds.</comment>
     <value>Problem starting the plugin '{0}'. {1}</value>
     <comment>0 - plugin path, 1 - error message </comment>
   </data>
-  <data name="PushCommandSkipDuplicateAdvertise" xml:space="preserve">
-    <value>See help for push option to automatically skip duplicates.</value>
+  <data name="PushCommandSkipDuplicateAdvertiseNuGetExe" xml:space="preserve">
+    <value>To skip already published packages, use &lt;command&gt; -SkipDuplicate</value>
```
Comment (loic-sharma) [inline src/NuGet.Core/NuGet.Protocol/Strings.resx]: Thoughts on this?

```suggestion
    <value>To skip an already published packages, use the -SkipDuplicate &lt;command&gt;</value>
```

Also, is command the right term? Isn't something like `--a` called a CLI option?

File: src/NuGet.Core/NuGet.Protocol/Strings.resx [hunk 7]
```diff
@@ -531,4 +531,7 @@ The "ms" should be localized to the abbreviation for milliseconds.</comment>
 {1} is the number of seconds
 {2} is the plugin process exit code</comment>
   </data>
-</root>
\ No newline at end of file
+  <data name="PushCommandSkipDuplicateAdvertiseDotnetExe" xml:space="preserve">
+    <value>To skip already published packages, use &lt;command&gt; --skip-duplicate</value>
```
Comment (loic-sharma) [inline src/NuGet.Core/NuGet.Protocol/Strings.resx]: Thoughts on this? See comment above

```suggestion
    <value>To skip an already published packages, use the --skip-duplicate &lt;command&gt;</value>
```

File: src/NuGet.Core/NuGet.Protocol/Strings.resx (hunk 6 above)
Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Protocol/Strings.resx]: "an" is not grammatically correct in the sentence. Are you wanting to change it to a singular package for some reason?

Mine says to use your command, followed by an option (<command> -SkipDuplicate). Your rewrite is not using the right term, as it's calling --a a command.

Comment (loic-sharma) [reply src/NuGet.Core/NuGet.Protocol/Strings.resx]: > "an" is not grammatically correct in the sentence. Are you wanting to change it to a singular package for some reason?

I misread the message 🤦‍♂️, `an ` is correct.

> Mine says to use your command, followed by an option. Your rewrite is not using the right term, as it's calling --a a command.

Ah I see. That's really confusing. Is this the pattern used by nuget.exe? I personally prefer the language `use the -SkipDuplicate option`.

Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Protocol/Strings.resx]: All CLIs I've ever seen use the pattern 

> < command> <option(s)>

. There's nothing unusual about that. If you want to reword it, you can propose that, but -SkipDuplicate never goes in front of the command.

Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Protocol/Strings.resx]: Changed `<command> -SkipDuplicate`  to `the option -SkipDuplicate`.

**Review by zivkan — APPROVED** [review]
